File: BsBacktraceLogger/ViewController.m
//
//  ViewController.m
//  KtBacktraceLogger
//
//  Created by 张星宇 on 16/8/26.
//  Copyright © 2016年 bestswifter. All rights reserved.
//



#import "ViewController.h"
#import "BsBacktraceLogger.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)foo {
    [self bar];
}

- (void)bar {
    while (true) {
        ;
    }
}

- (void)viewDidLoad {
    [super viewDidLoad];
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
        BSLOG_MAIN  // 打印主线程调用栈
        // 调用 [BsBacktraceLogger backtraceOfCurrentThread] 这一系列的方法可以获取字符串，然后选择上传服务器或者其他处理。
    });
    [self foo];
}

@end

File: Examples/ViewController.swift
//
//  ViewController.swift
//  Examples
//
//  Created by Ruswan Efendi on 29/03/22.
//  Copyright © 2022 Ruswan Efendi. All rights reserved.
//

import UIKit
import BSBacktraceLogger

class ViewController: UIViewController {

	override func viewDidLoad() {
		super.viewDidLoad()
		
		runPingThread()
		runLoop()
	}

	private func runPingThread() {
		
		let thread = Thread {
			while(true) {
				
				guard let backtrace = BSLogger.bs_backtraceOfMainThread() else {
					return
				}
				
				let traces = backtrace.split(separator: "\n")
				
				let parsedTraces = traces.compactMap({ trace -> (symbol: String, file: String, line: Int)? in
					let contents = trace.split(separator: " ")
					
					if contents.count < 5 {
						return nil
					}
					
					let fileName = String(contents[0])
					let symbol = contents[2..<(contents.count - 2)].joined(separator: " ")
					let line = (Int(contents[contents.count - 1]) ?? 0)
					
					return (symbol: symbol, file: fileName, line: line)
				})
				
				dump(parsedTraces)
			}
		}
		
		thread.start()
	}
	
	private func runLoop() {
		while(true) {
			Thread.sleep(forTimeInterval: 2)
		}
	}
}

